enterprise/server/backends/authdb/authdb.go

```diff
@@ -37,23 +42,86 @@ const (
 	apiKeyEncryptionBackfillBatchSize = 100
 )
 
+const (
+	// Maximum number of entries in API Key -> Group cache.
+	apiKeyGroupCacheSize = 10_000
+)
+
 var (
 	userOwnedKeysEnabled = flag.Bool("app.user_owned_keys_enabled", false, "If true, enable user-owned API keys.")
+	apiKeyGroupCacheTTL  = flag.Duration("auth.api_key_group_cache_ttl", 5*time.Minute, "TTL for API Key to Group caching. Set to '0' to disable cache.")
 	apiKeyEncryptionKey  = flagutil.New("auth.api_key_encryption.key", "", "Base64-encoded 256-bit encryption key for API keys.", flagutil.SecretTag)
 	encryptNewKeys       = flag.Bool("auth.api_key_encryption.encrypt_new_keys", false, "If enabled, all new API keys will be written in an encrypted format.")
 	encryptOldKeys       = flag.Bool("auth.api_key_encryption.encrypt_old_keys", false, "If enabled, all existing unencrypted keys will be encrypted on startup. The unencrypted keys will remain in the database and will need to be cleared manually after verifying the success of the migration.")
 )
 
+type apiKeyGroupCacheEntry struct {
+	data         interfaces.APIKeyGroup
+	expiresAfter time.Time
+}
+
+// apiKeyGroupCache is a cache for API Key -> Group lookups. A single Bazel invocation
+// can generate large bursts of RPCs, each of which needs to be authed.
+// There's no need to go to the database for every single request as this data
+// rarely changes.
+type apiKeyGroupCache struct {
+	// Note that even though we base this off an LRU cache, every entry has a hard expiration
+	// time to force a refresh of the underlying data.
+	lru interfaces.LRU[*apiKeyGroupCacheEntry]
+	ttl time.Duration
+	mu  sync.Mutex
+}
+
+func newAPIKeyGroupCache() (*apiKeyGroupCache, error) {
+	config := &lru.Config[*apiKeyGroupCacheEntry]{
+		MaxSize: apiKeyGroupCacheSize,
+		SizeFn:  func(v *apiKeyGroupCacheEntry) int64 { return 1 },
+	}
+	lru, err := lru.NewLRU[*apiKeyGroupCacheEntry](config)
+	if err != nil {
+		return nil, status.InternalErrorf("error initializing API Key -> Group cache: %v", err)
+	}
+	return &apiKeyGroupCache{lru: lru, ttl: *apiKeyGroupCacheTTL}, nil
+}
+
+func (c *apiKeyGroupCache) Get(apiKey string) (akg interfaces.APIKeyGroup, ok bool) {
+	c.mu.Lock()
+	entry, ok := c.lru.Get(apiKey)
+	c.mu.Unlock()
+	if !ok {
+		return nil, ok
+	}
+	if time.Now().After(entry.expiresAfter) {
+		return nil, false
+	}
+	return entry.data, true
+}
+
+func (c *apiKeyGroupCache) Add(apiKey string, apiKeyGroup interfaces.APIKeyGroup) {
+	c.mu.Lock()
+	c.lru.Add(apiKey, &apiKeyGroupCacheEntry{data: apiKeyGroup, expiresAfter: time.Now().Add(c.ttl)})
+	c.mu.Unlock()
+}
+
 type AuthDB struct {
 	env environment.Env
 	h   interfaces.DBHandle
 
+	apiKeyGroupCache *apiKeyGroupCache
+
 	// Nil if API key encryption is not enabled.
 	apiKeyEncryptionKey []byte
 }
 
 func NewAuthDB(env environment.Env, h interfaces.DBHandle) (*AuthDB, error) {
 	adb := &AuthDB{env: env, h: h}
+	if *apiKeyGroupCacheTTL != time.Duration(0) {
```

Comment:
if apiKeyGroupCacheTTL is a time.Duration you should just be able to compare it to 0?

Reply:
done